File: mRemoteV1/Config/Putty/Provider.vb

Imports mRemote3G.Connection
Imports mRemote3G.Connection.PuttySession
Imports mRemote3G.Images
Imports mRemote3G.Root.PuttySessions
Imports mRemote3G.Tree

Namespace Config.Putty
    Public MustInherit Class Provider

#Region "Public Methods"

        Private _rootTreeNode As TreeNode

        Public ReadOnly Property RootTreeNode As TreeNode
            Get
                If _rootTreeNode Is Nothing Then _rootTreeNode = CreateRootTreeNode()
                Return _rootTreeNode
            End Get
        End Property

        Private _rootInfo As PuttySessionsInfo

        Public ReadOnly Property RootInfo As PuttySessionsInfo
            Get
                If _rootInfo Is Nothing Then _rootInfo = CreateRootInfo()
                Return _rootInfo
            End Get
        End Property

        Public MustOverride Function GetSessionNames(Optional ByVal raw As Boolean = False) As String()
        Public MustOverride Function GetSession(sessionName As String) As PuttyInfo

        Public Overridable Function GetSessions() As PuttyInfo()
            Dim sessionList As New List(Of PuttyInfo)
            Dim sessionInfo As Info
            For Each sessionName As String In GetSessionNames(True)
                sessionInfo = GetSession(sessionName)
                If sessionInfo Is Nothing OrElse String.IsNullOrEmpty(sessionInfo.Hostname) Then Continue For
                sessionList.Add(sessionInfo)
            Next
            Return sessionList.ToArray()
        End Function

        Public Overridable Sub StartWatcher()
        End Sub

        Public Overridable Sub StopWatcher()
        End Sub

#End Region

#Region "Public Events"

        Public Event SessionChanged(sender As Object, e As SessionChangedEventArgs)

#End Region

#Region "Public Classes"

        Public Class SessionChangedEventArgs
            Inherits EventArgs
        End Class

#End Region

#Region "Protected Methods"

        Private Delegate Function CreateRootTreeNodeDelegate() As TreeNode

        Protected Overridable Function CreateRootTreeNode() As TreeNode
            Dim treeView As TreeView = Node.TreeView
            If treeView Is Nothing Then Return Nothing
            If treeView.InvokeRequired Then
                Return treeView.Invoke(New CreateRootTreeNodeDelegate(AddressOf CreateRootTreeNode))
            End If

            Dim newTreeNode As New TreeNode
            RootInfo.TreeNode = newTreeNode

            newTreeNode.Name = _rootInfo.Name
            newTreeNode.Text = _rootInfo.Name
            newTreeNode.Tag = _rootInfo
            newTreeNode.ImageIndex = Enums.TreeImage.PuttySessions
            newTreeNode.SelectedImageIndex = Enums.TreeImage.PuttySessions

            Return newTreeNode
        End Function

        Protected Overridable Function CreateRootInfo() As PuttySessionsInfo
            Dim newRootInfo As New PuttySessionsInfo

            If String.IsNullOrEmpty(My.Settings.PuttySavedSessionsName) Then
                newRootInfo.Name = Language.Language.strPuttySavedSessionsRootName
            Else
                newRootInfo.Name = My.Settings.PuttySavedSessionsName
            End If

            If String.IsNullOrEmpty(My.Settings.PuttySavedSessionsPanel) Then
                newRootInfo.Panel = Language.Language.strGeneral
            Else
                newRootInfo.Panel = My.Settings.PuttySavedSessionsPanel
            End If

            Return newRootInfo
        End Function

        Protected Overridable Sub OnSessionChanged(e As SessionChangedEventArgs)
            RaiseEvent SessionChanged(Me, New SessionChangedEventArgs())
        End Sub

#End Region
    End Class
End Namespace

File: mRemoteV1/CS/Config/Putty/Provider.cs
// VBConversions Note: VB project level imports
using System.Collections.Generic;
using System;
using AxWFICALib;
using System.Drawing;
using System.Diagnostics;
using System.Data;
using AxMSTSCLib;
using Microsoft.VisualBasic;
using System.Collections;
using System.Windows.Forms;
// End of VB project level imports

using mRemoteNG.My;


namespace mRemoteNG.Config.Putty
{
	public abstract class Provider
		{
#region Public Methods
			private TreeNode _rootTreeNode;
public TreeNode RootTreeNode
			{
				get
				{
					if (_rootTreeNode == null)
					{
						_rootTreeNode = CreateRootTreeNode();
					}
					return _rootTreeNode;
				}
			}
			
			private Root.PuttySessions.Info _rootInfo;
public Root.PuttySessions.Info RootInfo
			{
				get
				{
					if (_rootInfo == null)
					{
						_rootInfo = CreateRootInfo();
					}
					return _rootInfo;
				}
			}
			
			public abstract string[] GetSessionNames(bool raw = false);
			public abstract Connection.PuttySession.Info GetSession(string sessionName);
			
			public virtual Connection.PuttySession.Info[] GetSessions()
			{
				List<Connection.PuttySession.Info> sessionList = new List<Connection.PuttySession.Info>();
				Connection.Info sessionInfo = default(Connection.Info);
				foreach (string sessionName in GetSessionNames(true))
				{
					sessionInfo = GetSession(sessionName);
					if (sessionInfo == null || string.IsNullOrEmpty(sessionInfo.Hostname))
					{
						continue;
					}
					sessionList.Add(sessionInfo);
				}
				return sessionList.ToArray();
			}
			
			public virtual void StartWatcher()
			{
				
			}
			
			public virtual void StopWatcher()
			{
				
			}
#endregion
			
#region Public Events
			public delegate void SessionChangedEventHandler(object sender, SessionChangedEventArgs e);
			private SessionChangedEventHandler SessionChangedEvent;
			
			public event SessionChangedEventHandler SessionChanged
			{
				add
				{
					SessionChangedEvent = (SessionChangedEventHandler) System.Delegate.Combine(SessionChangedEvent, value);
				}
				remove
				{
					SessionChangedEvent = (SessionChangedEventHandler) System.Delegate.Remove(SessionChangedEvent, value);
				}
			}
			
#endregion
			
#region Public Classes
			public class SessionChangedEventArgs : EventArgs
			{
			}
#endregion
			
#region Protected Methods
			private delegate TreeNode  CreateRootTreeNodeDelegate();
			protected virtual TreeNode CreateRootTreeNode()
			{
				TreeView treeView = Tree.Node.TreeView;
				if (treeView == null)
				{
					return null;
				}
				if (treeView.InvokeRequired)
				{
					return treeView.Invoke(new CreateRootTreeNodeDelegate(CreateRootTreeNode));
				}
				
				TreeNode newTreeNode = new TreeNode();
				RootInfo.TreeNode = newTreeNode;
				
				newTreeNode.Name = _rootInfo.Name;
				newTreeNode.Text = _rootInfo.Name;
				newTreeNode.Tag = _rootInfo;
				newTreeNode.ImageIndex = Images.Enums.TreeImage.PuttySessions;
				newTreeNode.SelectedImageIndex = Images.Enums.TreeImage.PuttySessions;
				
				return newTreeNode;
			}
			
			protected virtual Root.PuttySessions.Info CreateRootInfo()
			{
				Root.PuttySessions.Info newRootInfo = new Root.PuttySessions.Info();
				
				if (string.IsNullOrEmpty(System.Convert.ToString(My.Settings.Default.PuttySavedSessionsName)))
				{
					newRootInfo.Name = Language.strPuttySavedSessionsRootName;
				}
				else
				{
					newRootInfo.Name = System.Convert.ToString(My.Settings.Default.PuttySavedSessionsName);
				}
				
				if (string.IsNullOrEmpty(System.Convert.ToString(My.Settings.Default.PuttySavedSessionsPanel)))
				{
					newRootInfo.Panel = Language.strGeneral;
				}
				else
				{
					newRootInfo.Panel = System.Convert.ToString(My.Settings.Default.PuttySavedSessionsPanel);
				}
				
				return newRootInfo;
			}
			
			protected virtual void OnSessionChanged(SessionChangedEventArgs e)
			{
				if (SessionChangedEvent != null)
					SessionChangedEvent(this, new SessionChangedEventArgs());
			}
#endregion
		}
}
